WordPressKit/WordPressComOAuthClient.swift

```diff
@@ -1,15 +1,106 @@
 import Alamofire
 
-@objc public enum WordPressComOAuthError: Int {
-    case unknown
-    case invalidClient
-    case unsupportedGrantType
-    case invalidRequest
-    case needsMultifactorCode
-    case invalidOneTimePassword
-    case socialLoginExistingUserUnconnected
-    case invalidTwoStepCode
-    case unknownUser
+/// Errors returned by `WordPressComOAuthClient`.
+public enum WordPressComOAuthError: Error {
+    /// Can't encode the request arguments into a valid HTTP request. This is a programming error.
+    case requestEncodingFailure
+    /// Error occured in the HTTP connection.
+    case connection(URLError)
+    /// The caller didn't pass a closure to handle MFA authentication request. This is a programming error.
+    case missingMultiFactorHandler
+    /// WordPress.com rejects the authentication request and returns an error. See `AuthenticationFailure` for error details.
+    case authenticationFailure(AuthenticationFailure)
+    /// WordPress.com returns an HTTP response that WordPressKit can't parse.
+    case unparsableResponse(response: HTTPURLResponse?, body: Data?)
+    /// Other error occured.
+    case unknown(underlyingError: Error)
+
+    public var authenticationFailureKind: AuthenticationFailure.Kind? {
+        if case let .authenticationFailure(authenticationFailure) = self {
```

Comment:
Oh, interesting. I usually would have written `case .authenticationFailure(let authenticationFailure)` nice to see this works, too.